Bug in loading directory inside docker container

I'm running the following series of commands
```
kpt pkg get https://github.com/kubernetes-sigs/kustomize/examples/helloWorld helloWorld
kpt fn source helloWorld |
  kpt fn run --mount type=bind,src="$(pwd)",dst=/source,rw=true --image gcr.io/kpt-functions/kustomize-build:"${TAG}" -- path=/source/helloWorld
```
I expect this to run `kustomize build /source/helloWorld` where the build path is mounted into the docker container. But I get a permission error from [here](https://github.com/kubernetes-sigs/kustomize/blob/master/api/loader/loader.go) even though I have full permission to the helloWorld folder. 
```
fetching package examples/helloWorld from https://github.com/kubernetes-sigs/kustomize to helloWorld
[ERROR] Kustomize build command results in error: Error: Error creating new loader with git: uri looks like abs path: /source/helloWorld, dir: evalsymlink failure on '/source/helloWorld' : lstat /source/helloWorld: permission denied, get: invalid source string: /source/helloWorld
```

Hey @aodinokov , I was building kpt from source before. Now I'm on the latest version and I modified the command to get around load restrictor but still seeing an error about not finding any kustomization.
```
$ kpt fn source helloWorld | kpt fn run --mount type=bind,src="$(pwd)",dst=/source,rw=true --image gcr.io/kpt-functions/kustomize-build:latest -- path=/source/helloWorld
[ERROR] Kustomize build command results in error: Error: Error creating new loader with git: uri looks like abs path: /source/helloWorld, dir: evalsymlink failure on '/source/helloWorld' : lstat /source/helloWorld: permission denied, get: invalid source string: /source/helloWorld

Error: exit status 1
```
```
$ kpt fn source helloWorld | kpt fn run --mount type=bind,src="$(pwd)/helloWorld",target=/source --image gcr.io/kpt-functions/kustomize-build:latest -- path=/source
[ERROR] Kustomize build command results in error: Error: unable to find one of 'kustomization.yaml', 'kustomization.yml' or 'Kustomization' in directory '/source'

Error: exit status 1
```

--- a/docs/reference/fn/index.xml
+++ b/docs/reference/fn/index.xml
@@ -107,7 +107,22 @@ kpt sink DIR/
 &lt;code&gt;ConfigMap&lt;/code&gt; containing &lt;code&gt;data: {foo: bar}&lt;/code&gt;. This is used to parameterize the behavior
 of the function:&lt;/p&gt;
 &lt;div class=&#34;highlight&#34;&gt;&lt;pre style=&#34;background-color:#f8f8f8;-moz-tab-size:4;-o-tab-size:4;tab-size:4&#34;&gt;&lt;code class=&#34;language-sh&#34; data-lang=&#34;sh&#34;&gt;kpt fn run DIR/ --image gcr.io/example.com/my-fn -- &lt;span style=&#34;color:#000&#34;&gt;foo&lt;/span&gt;&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;bar
-&lt;/code&gt;&lt;/pre&gt;&lt;/div&gt;&lt;h2 id=&#34;declaratively-run-one-or-more-functions&#34;&gt;Declaratively run one or more functions&lt;/h2&gt;
+&lt;/code&gt;&lt;/pre&gt;&lt;/div&gt;&lt;h3 id=&#34;caveats-to-running-imperatively&#34;&gt;Caveats to running imperatively&lt;/h3&gt;
+&lt;h4 id=&#34;arguments-interpreted-as-flags&#34;&gt;Arguments interpreted as flags&lt;/h4&gt;
+&lt;p&gt;Some functions like &lt;code&gt;helm-template&lt;/code&gt;, &lt;code&gt;istioctl-analyze&lt;/code&gt;, and &lt;code&gt;kustomize-build&lt;/code&gt; take
+arbitrary command line flags as arguments. Passing arguments such as &lt;code&gt;--use-kube=false&lt;/code&gt;
+imperatively results in parsing issues. See more details
+&lt;a href=&#34;https://github.com/GoogleContainerTools/kpt/issues/823/&#34;&gt;here&lt;/a&gt; and
+&lt;a href=&#34;https://github.com/GoogleContainerTools/kpt/issues/824/&#34;&gt;here&lt;/a&gt;.&lt;/p&gt;
+&lt;p&gt;When passing flags as arguments, it&amp;rsquo;s recommended to run functions declaratively so they
+can be run across various orchestrators.&lt;/p&gt;
+&lt;h4 id=&#34;functions-expecting-spec-field&#34;&gt;Functions expecting spec field&lt;/h4&gt;
+&lt;p&gt;&lt;code&gt;kpt fn run&lt;/code&gt; provides any arguments passed imperatively to the container image in a
+&lt;code&gt;ConfigMap&lt;/code&gt; containing a &lt;code&gt;data&lt;/code&gt; field. Some config functions may expect arguemnts passed in
+a &lt;code&gt;spec&lt;/code&gt; field instead. It&amp;rsquo;s recommended to run such functions declaratively by passing a
+&lt;code&gt;ConfigMap&lt;/code&gt; with a &lt;code&gt;spec&lt;/code&gt; field. See more details
+&lt;a href=&#34;https://github.com/GoogleContainerTools/kpt/issues/757&#34;&gt;here&lt;/a&gt;.&lt;/p&gt;
+&lt;h2 id=&#34;declaratively-run-one-or-more-functions&#34;&gt;Declaratively run one or more functions&lt;/h2&gt;
 &lt;p&gt;Functions and their input configuration may be declared in files rather than directly
 on the command line.&lt;/p&gt;
 &lt;p&gt;&lt;strong&gt;Example:&lt;/strong&gt; This is equivalent to the preceding example:&lt;/p&gt;
@@ -198,7 +213,22 @@ container &lt;code&gt;STDERR&lt;/code&gt;.&lt;/p&gt;
 exiting with a non-zero status code. Users may want to store these results separately from configuration
 files. Kpt provides the &lt;code&gt;--results-dir&lt;/code&gt; flag for users to specify a destination to write results to.&lt;/p&gt;
 &lt;div class=&#34;highlight&#34;&gt;&lt;pre style=&#34;background-color:#f8f8f8;-moz-tab-size:4;-o-tab-size:4;tab-size:4&#34;&gt;&lt;code class=&#34;language-sh&#34; data-lang=&#34;sh&#34;&gt;kpt fn run DIR --results-dir RESULTS_DIR --fn-path FUNCTIONS_DIR/
-&lt;/code&gt;&lt;/pre&gt;&lt;/div&gt;&lt;h2 id=&#34;next-steps&#34;&gt;Next Steps&lt;/h2&gt;
+&lt;/code&gt;&lt;/pre&gt;&lt;/div&gt;&lt;h2 id=&#34;mounting-directories&#34;&gt;Mounting Directories&lt;/h2&gt;
+&lt;p&gt;You can mount a local directory to the container image using the &lt;code&gt;--mount&lt;/code&gt; flag
+passing the same arguments as for &lt;code&gt;docker run&lt;/code&gt;. See the &lt;a href=&#34;https://docs.docker.com/engine/reference/commandline/run/#add-bind-mounts-or-volumes-using-the---mount-flag&#34;&gt;docker docs&lt;/a&gt; for more
+details on the &lt;code&gt;--mount&lt;/code&gt; flag.&lt;/p&gt;
+&lt;div class=&#34;highlight&#34;&gt;&lt;pre style=&#34;background-color:#f8f8f8;-moz-tab-size:4;-o-tab-size:4;tab-size:4&#34;&gt;&lt;code class=&#34;language-sh&#34; data-lang=&#34;sh&#34;&gt;kpt fn &lt;span style=&#34;color:#204a87&#34;&gt;source&lt;/span&gt; helloWorld &lt;span style=&#34;color:#000;font-weight:bold&#34;&gt;|&lt;/span&gt;
+  kpt fn run --mount &lt;span style=&#34;color:#000&#34;&gt;type&lt;/span&gt;&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;bind,src&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;&lt;span style=&#34;color:#4e9a06&#34;&gt;&amp;#34;&lt;/span&gt;&lt;span style=&#34;color:#204a87;font-weight:bold&#34;&gt;$(&lt;/span&gt;&lt;span style=&#34;color:#204a87&#34;&gt;pwd&lt;/span&gt;&lt;span style=&#34;color:#204a87;font-weight:bold&#34;&gt;)&lt;/span&gt;&lt;span style=&#34;color:#4e9a06&#34;&gt;/helloWorld&amp;#34;&lt;/span&gt;,dst&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;/source --image gcr.io/kpt-functions/kustomize-build -- &lt;span style=&#34;color:#000&#34;&gt;path&lt;/span&gt;&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;/source &lt;span style=&#34;color:#000;font-weight:bold&#34;&gt;|&lt;/span&gt;
+  kpt fn sink .
+&lt;/code&gt;&lt;/pre&gt;&lt;/div&gt;&lt;p&gt;Kpt provides mounts as read-only by default. Explicitly pass &lt;code&gt;rw=true&lt;/code&gt; to
+provide a read-write mount. See the example command below.&lt;/p&gt;
+&lt;div class=&#34;highlight&#34;&gt;&lt;pre style=&#34;background-color:#f8f8f8;-moz-tab-size:4;-o-tab-size:4;tab-size:4&#34;&gt;&lt;code class=&#34;language-sh&#34; data-lang=&#34;sh&#34;&gt;kpt fn &lt;span style=&#34;color:#204a87&#34;&gt;source&lt;/span&gt; helloWorld &lt;span style=&#34;color:#000;font-weight:bold&#34;&gt;|&lt;/span&gt;
+  kpt fn run --mount &lt;span style=&#34;color:#000&#34;&gt;type&lt;/span&gt;&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;bind,src&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;&lt;span style=&#34;color:#4e9a06&#34;&gt;&amp;#34;&lt;/span&gt;&lt;span style=&#34;color:#204a87;font-weight:bold&#34;&gt;$(&lt;/span&gt;&lt;span style=&#34;color:#204a87&#34;&gt;pwd&lt;/span&gt;&lt;span style=&#34;color:#204a87;font-weight:bold&#34;&gt;)&lt;/span&gt;&lt;span style=&#34;color:#4e9a06&#34;&gt;/helloWorld&amp;#34;&lt;/span&gt;,dst&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;/source,rw&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;&lt;span style=&#34;color:#204a87&#34;&gt;true&lt;/span&gt; --image gcr.io/kpt-functions/kustomize-build -- &lt;span style=&#34;color:#000&#34;&gt;path&lt;/span&gt;&lt;span style=&#34;color:#ce5c00;font-weight:bold&#34;&gt;=&lt;/span&gt;/source &lt;span style=&#34;color:#000;font-weight:bold&#34;&gt;|&lt;/span&gt;
+  kpt fn sink .
+&lt;/code&gt;&lt;/pre&gt;&lt;/div&gt;&lt;p&gt;Depending on the container image, the configuration function may not have
+permissions to access mounted volumes. Check how the function is running inside
+the container in case of permissions issues.&lt;/p&gt;
+&lt;h2 id=&#34;next-steps&#34;&gt;Next Steps&lt;/h2&gt;
 &lt;ul&gt;
 &lt;li&gt;See more examples of functions in the &lt;a href=&#34;../../../guides/consumer/function/catalog/&#34;&gt;functions catalog&lt;/a&gt;.&lt;/li&gt;
 &lt;li&gt;Get a quickstart on writing functions from the &lt;a href=&#34;../../../guides/producer/functions/&#34;&gt;function producer docs&lt;/a&gt;.&lt;/li&gt;

--- a/docs/reference/fn/run/index.html
+++ b/docs/reference/fn/run/index.html
@@ -929,13 +929,18 @@ if (!doNotTrack) {
 <nav id="TableOfContents">
   <ul>
     <li><a href="#examples">Examples</a></li>
-    <li><a href="#imperatively-run-a-single-function">Imperatively run a single function</a></li>
+    <li><a href="#imperatively-run-a-single-function">Imperatively run a single function</a>
+      <ul>
+        <li><a href="#caveats-to-running-imperatively">Caveats to running imperatively</a></li>
+      </ul>
+    </li>
     <li><a href="#declaratively-run-one-or-more-functions">Declaratively run one or more functions</a></li>
     <li><a href="#scoping-rules">Scoping Rules</a></li>
     <li><a href="#declaring-multiple-functions">Declaring Multiple Functions</a></li>
     <li><a href="#custom-functionconfig">Custom <code>functionConfig</code></a></li>
     <li><a href="#synopsis">Synopsis</a></li>
     <li><a href="#structured-results">Structured Results</a></li>
+    <li><a href="#mounting-directories">Mounting Directories</a></li>
     <li><a href="#next-steps">Next Steps</a></li>
   </ul>
 </nav>
@@ -1021,7 +1026,22 @@ kpt sink DIR/
 <code>ConfigMap</code> containing <code>data: {foo: bar}</code>. This is used to parameterize the behavior
 of the function:</p>
 <div class="highlight"><pre style="background-color:#f8f8f8;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">kpt fn run DIR/ --image gcr.io/example.com/my-fn -- <span style="color:#000">foo</span><span style="color:#ce5c00;font-weight:bold">=</span>bar
-</code></pre></div><h2 id="declaratively-run-one-or-more-functions">Declaratively run one or more functions</h2>
+</code></pre></div><h3 id="caveats-to-running-imperatively">Caveats to running imperatively</h3>
+<h4 id="arguments-interpreted-as-flags">Arguments interpreted as flags</h4>
+<p>Some functions like <code>helm-template</code>, <code>istioctl-analyze</code>, and <code>kustomize-build</code> take
+arbitrary command line flags as arguments. Passing arguments such as <code>--use-kube=false</code>
+imperatively results in parsing issues. See more details
+<a href="https://github.com/GoogleContainerTools/kpt/issues/823/">here</a> and
+<a href="https://github.com/GoogleContainerTools/kpt/issues/824/">here</a>.</p>
+<p>When passing flags as arguments, it&rsquo;s recommended to run functions declaratively so they
+can be run across various orchestrators.</p>
+<h4 id="functions-expecting-spec-field">Functions expecting spec field</h4>
+<p><code>kpt fn run</code> provides any arguments passed imperatively to the container image in a
+<code>ConfigMap</code> containing a <code>data</code> field. Some config functions may expect arguemnts passed in
+a <code>spec</code> field instead. It&rsquo;s recommended to run such functions declaratively by passing a
+<code>ConfigMap</code> with a <code>spec</code> field. See more details
+<a href="https://github.com/GoogleContainerTools/kpt/issues/757">here</a>.</p>
+<h2 id="declaratively-run-one-or-more-functions">Declaratively run one or more functions</h2>
 <p>Functions and their input configuration may be declared in files rather than directly
 on the command line.</p>
 <p><strong>Example:</strong> This is equivalent to the preceding example:</p>
@@ -1112,7 +1132,22 @@ container <code>STDERR</code>.</p>
 exiting with a non-zero status code. Users may want to store these results separately from configuration
 files. Kpt provides the <code>--results-dir</code> flag for users to specify a destination to write results to.</p>
 <div class="highlight"><pre style="background-color:#f8f8f8;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">kpt fn run DIR --results-dir RESULTS_DIR --fn-path FUNCTIONS_DIR/
-</code></pre></div><h2 id="next-steps">Next Steps</h2>
+</code></pre></div><h2 id="mounting-directories">Mounting Directories</h2>
+<p>You can mount a local directory to the container image using the <code>--mount</code> flag
+passing the same arguments as for <code>docker run</code>. See the <a href="https://docs.docker.com/engine/reference/commandline/run/#add-bind-mounts-or-volumes-using-the---mount-flag">docker docs</a> for more
+details on the <code>--mount</code> flag.</p>
+<div class="highlight"><pre style="background-color:#f8f8f8;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">kpt fn <span style="color:#204a87">source</span> helloWorld <span style="color:#000;font-weight:bold">|</span>
+  kpt fn run --mount <span style="color:#000">type</span><span style="color:#ce5c00;font-weight:bold">=</span>bind,src<span style="color:#ce5c00;font-weight:bold">=</span><span style="color:#4e9a06">&#34;</span><span style="color:#204a87;font-weight:bold">$(</span><span style="color:#204a87">pwd</span><span style="color:#204a87;font-weight:bold">)</span><span style="color:#4e9a06">/helloWorld&#34;</span>,dst<span style="color:#ce5c00;font-weight:bold">=</span>/source --image gcr.io/kpt-functions/kustomize-build -- <span style="color:#000">path</span><span style="color:#ce5c00;font-weight:bold">=</span>/source <span style="color:#000;font-weight:bold">|</span>
+  kpt fn sink .
+</code></pre></div><p>Kpt provides mounts as read-only by default. Explicitly pass <code>rw=true</code> to
+provide a read-write mount. See the example command below.</p>
+<div class="highlight"><pre style="background-color:#f8f8f8;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">kpt fn <span style="color:#204a87">source</span> helloWorld <span style="color:#000;font-weight:bold">|</span>
+  kpt fn run --mount <span style="color:#000">type</span><span style="color:#ce5c00;font-weight:bold">=</span>bind,src<span style="color:#ce5c00;font-weight:bold">=</span><span style="color:#4e9a06">&#34;</span><span style="color:#204a87;font-weight:bold">$(</span><span style="color:#204a87">pwd</span><span style="color:#204a87;font-weight:bold">)</span><span style="color:#4e9a06">/helloWorld&#34;</span>,dst<span style="color:#ce5c00;font-weight:bold">=</span>/source,rw<span style="color:#ce5c00;font-weight:bold">=</span><span style="color:#204a87">true</span> --image gcr.io/kpt-functions/kustomize-build -- <span style="color:#000">path</span><span style="color:#ce5c00;font-weight:bold">=</span>/source <span style="color:#000;font-weight:bold">|</span>
+  kpt fn sink .
+</code></pre></div><p>Depending on the container image, the configuration function may not have
+permissions to access mounted volumes. Check how the function is running inside
+the container in case of permissions issues.</p>
+<h2 id="next-steps">Next Steps</h2>
 <ul>
 <li>See more examples of functions in the <a href="../../../guides/consumer/function/catalog/">functions catalog</a>.</li>
 <li>Get a quickstart on writing functions from the <a href="../../../guides/producer/functions/">function producer docs</a>.</li>
@@ -1193,7 +1228,7 @@ files. Kpt provides the <code>--results-dir</code> flag for users to specify a d
 
 	
 	
-	<div class="text-muted mt-5 pt-3 border-top">Last modified July 7, 2020: <a  href="https://github.com/GoogleContainerTools/kpt/commit/c78d6c566a905ab2b7466a85c70435cc8d3d3ed2">Docs: Fix W3 Link Checker results (c78d6c56)</a>
+	<div class="text-muted mt-5 pt-3 border-top">Last modified July 27, 2020: <a  href="https://github.com/GoogleContainerTools/kpt/commit/f28edb096721821a1fee9158c7887487b059ce13">Docs: Mounting volumes to kpt fn run (f28edb09)</a>
 </div>
 </div>
 

--- a/docs/sitemap.xml
+++ b/docs/sitemap.xml
@@ -414,7 +414,7 @@
   
   <url>
     <loc>https://googlecontainertools.github.io/kpt/reference/fn/run/</loc>
-    <lastmod>2020-07-07T13:09:02-04:00</lastmod>
+    <lastmod>2020-07-27T16:30:57-07:00</lastmod>
   </url>
   
   <url>

--- a/site/content/en/reference/fn/run/_index.md
+++ b/site/content/en/reference/fn/run/_index.md
@@ -72,6 +72,27 @@ of the function:
 kpt fn run DIR/ --image gcr.io/example.com/my-fn -- foo=bar
 ```
 
+### Caveats to running imperatively
+
+#### Arguments interpreted as flags
+
+Some functions like `helm-template`, `istioctl-analyze`, and `kustomize-build` take
+arbitrary command line flags as arguments. Passing arguments such as `--use-kube=false`
+imperatively results in parsing issues. See more details
+[here](https://github.com/GoogleContainerTools/kpt/issues/823/) and
+[here](https://github.com/GoogleContainerTools/kpt/issues/824/).
+
+When passing flags as arguments, it's recommended to run functions declaratively so they
+can be run across various orchestrators.
+
+#### Functions expecting spec field
+
+`kpt fn run` provides any arguments passed imperatively to the container image in a
+`ConfigMap` containing a `data` field. Some config functions may expect arguemnts passed in
+a `spec` field instead. It's recommended to run such functions declaratively by passing a
+`ConfigMap` with a `spec` field. See more details
+[here](https://github.com/GoogleContainerTools/kpt/issues/757).
+
 ## Declaratively run one or more functions
 
 Functions and their input configuration may be declared in files rather than directly
@@ -211,6 +232,31 @@ files. Kpt provides the `--results-dir` flag for users to specify a destination 
 kpt fn run DIR --results-dir RESULTS_DIR --fn-path FUNCTIONS_DIR/
 ```
 
+## Mounting Directories
+
+You can mount a local directory to the container image using the `--mount` flag
+passing the same arguments as for `docker run`. See the [docker docs] for more
+details on the `--mount` flag.
+
+```sh
+kpt fn source helloWorld |
+  kpt fn run --mount type=bind,src="$(pwd)/helloWorld",dst=/source --image gcr.io/kpt-functions/kustomize-build -- path=/source |
+  kpt fn sink .
+```
+
+Kpt provides mounts as read-only by default. Explicitly pass `rw=true` to
+provide a read-write mount. See the example command below.
+
+```sh
+kpt fn source helloWorld |
+  kpt fn run --mount type=bind,src="$(pwd)/helloWorld",dst=/source,rw=true --image gcr.io/kpt-functions/kustomize-build -- path=/source |
+  kpt fn sink .
+```
+
+Depending on the container image, the configuration function may not have
+permissions to access mounted volumes. Check how the function is running inside
+the container in case of permissions issues.
+
 ## Next Steps
 
 - See more examples of functions in the [functions catalog].
@@ -221,3 +267,4 @@ kpt fn run DIR --results-dir RESULTS_DIR --fn-path FUNCTIONS_DIR/
 [functions catalog]: ../../../guides/consumer/function/catalog/
 [function producer docs]: ../../../guides/producer/functions/
 [functions concepts]: ../../../concepts/functions/
+[docker docs]: https://docs.docker.com/engine/reference/commandline/run/#add-bind-mounts-or-volumes-using-the---mount-flag

